### PR TITLE
Flydsl mxfp8 quantize

### PR DESCRIPTION
**Pull request by zstreet87**
Summary

  Adds the FlyDSL backend for MXFP8 quantize kernels on AMD CDNA, mirroring the existing
  cuTeDSL surface so the dispatcher can route to either backend by hardware. Three new kernels
   (2d_1x32, 2d_32x1, 3d) ship with matching dispatcher entrypoints, custom-op registration,
  tests, and benchmarks.

  What's in this PR

  - 3 FlyDSL kernels
  (torchao/prototype/moe_training/kernels/mxfp8/flydsl_quantize_{2d_1x32,2d_32x1,3d}.py) —
  same I/O contract as the cutedsl peers, with FlyDSL-specific perf work already applied
  (multi-wave workgroups, buffer_load_dwordx2 widening, M-stacked WG layout).
  - Dispatcher integration — mxfp8_quantize_*_flydsl entrypoints in quant.py, registered as
  torch.library custom ops, gated on a _mxfp8_flydsl_kernels_available flag and wired into
  MXFP8Dim1CastKernelChoice.
  - API surface parity — kernel-module signatures accept the same kwargs as the cutedsl
  wrappers (stage_count, blocked_scale_output, offs, scale_block_n); unsupported values raise
  NotImplementedError at both the dispatcher and kernel-module boundaries.
  - 3D blocked scale output + scale_block_k=32 implemented (matches cutedsl 3D's full option
  set).
  - Tests — numerics tests for all 3 kernels mirror cutedsl coverage where divisibility
  allows; dispatcher and kernel-module raise contracts both asserted; custom-op registration
  checked.
  - Benchmarks — new bench_flydsl_quantize_2d_{1x32,32x1}.py mirror the cutedsl bench files
  (FLOOR-only, baseline = triton_to_mxfp8_dim{0,1}); bench_quantize_3d.py extended with a
  flydsl column and gated so the script runs on either backend.

  Known limitations (follow-up PRs)

  - 1x32 requires K % 2048 == 0 (one workgroup-iteration consumes wave_size × block_size).
  Supporting K=7168 needs tail-handling in the kernel — see ROCm/FlyDSL PR #433's
  compute_compile_constants + make_pingpong_kloop for the compile-time-loop +
  runtime-tail-guard pattern.
  - scaling_mode="rceil" not implemented. The 3-op exponent-roundup pattern is straightforward
   to port from ROCm/FlyDSL PR #433's
  tests/kernels/blockscale_gemm_test_utils.py:fp32_to_e8m0: extract exponent byte, add 1 iff
  mantissa is nonzero, clamp to [1, 254]. Required for correctness — FLOOR systematically
  biases scales down and causes FP8 saturation.
  - 2D blocked_scale_output not implemented (tcgen05-specific scale layout).
  - offs (token-group offsets) not yet wired into the 2D kernels.
  - 3D scale_block_n != 32 not implemented.

  All raise NotImplementedError with descriptive messages.

  Test plan

  - pytest test/prototype/moe_training/test_kernels.py -k "flydsl or amd_mx_3d_flydsl" — 458
  passing on gfx950
  - pytest test/prototype/moe_training/test_kernels.py::test_flydsl_dispatcher_rejects_unsuppo
  rted_options — contract test passes
  - pytest test/prototype/moe_training/test_kernels.py::test_flydsl_kernel_module_rejects_unsu
  pported_options — direct-import contract test passes
  - pytest test/prototype/moe_training/test_kernels.py::test_flydsl_custom_ops_registered —
  ops registered
  - python benchmarks/prototype/moe_training/mxfp8/bench_flydsl_quantize_2d_1x32.py — runs
  end-to-end on AMD
  - python benchmarks/prototype/moe_training/mxfp8/bench_flydsl_quantize_2d_32x1.py — runs
  end-to-end on AMD
  - python benchmarks/prototype/moe_training/mxfp8/bench_quantize_3d.py — runs on either
  backend; columns NaN'd on the absent one
  
   @vkuzo — measured on AMD Instinct MI355X (gfx950), HBM3E spec peak = 8.0 TB/s. Comparison is FlyDSL
   vs the next-best AMD path (triton: triton_to_mxfp8_dim0 for 2D 1x32, triton_to_mxfp8_dim1 for 2D
   32x1, per-expert triton_to_mxfp8_dim1 loop for 3D — there is no native 3D triton MXFP8 kernel on
   AMD, so the loop is the honest fallback). Methodology unchanged from the bench scripts in this PR;
   FLOOR scaling; bytes = read(bf16) + write(fp8) + write(scale_e8m0). % of peak is vs spec-sheet 8
   TB/s (not measured achievable peak, which is typically ~80–90% of spec — so % of achievable would be
    ~10–20 pts higher).
  
   2D 1x32 (bench_flydsl_quantize_2d_1x32.py)
  
   | shape (M, K)   | flydsl GB/s (%peak) | triton GB/s (%peak) | speedup |
   |----------------|---------------------|---------------------|---------|
   | (8192, 2048)   | 1746 (21.8%)        | 497 (6.2%)          | 3.52×   |
   | (32768, 2048)  | 3204 (40.1%)        | 1763 (22.0%)        | 1.82×   |
   | (131072, 2048) | 4157 (52.0%)        | 3360 (42.0%)        | 1.24×   |
  
   2D 32x1 (bench_flydsl_quantize_2d_32x1.py)
  
   | shape (M, K)   | flydsl GB/s (%peak) | triton GB/s (%peak) | speedup |
   |----------------|---------------------|---------------------|---------|
   | (8192, 2048)   | 1781 (22.3%)        | 518 (6.5%)          | 3.44×   |
   | (8192, 7168)   | 2478 (31.0%)        | 1638 (20.5%)        | 1.51×   |
   | (32768, 2048)  | 2289 (28.6%)        | 2171 (27.1%)        | 1.05×   |
   | (32768, 7168)  | 3328 (41.6%)        | 2908 (36.4%)        | 1.14×   |
   | (131072, 2048) | 2997 (37.5%)        | 3068 (38.4%)        | 0.98×   |
   | (131072, 7168) | 3772 (47.2%)        | 3083 (38.5%)        | 1.22×   |
  
   3D FLOOR, scale_block_k=1 (bench_quantize_3d.py; triton baseline
   is a per-expert triton_to_mxfp8_dim1 loop)
  
   | shape (E, N, K)  | flydsl GB/s (%peak) | triton GB/s (%peak) | speedup |
   |------------------|---------------------|---------------------|---------|
   | (1, 8192, 5120)  | 2428 (30.4%)        | 1352 (16.9%)        | 1.80×   |
   | (1, 7168, 2048)  | 1344 (16.8%)        | 461 (5.8%)          | 2.91×   |
   | (8, 8192, 5120)  | 2868 (35.8%)        | 1338 (16.7%)        | 2.14×   |
   | (8, 7168, 2048)  | 2433 (30.4%)        | 478 (6.0%)          | 5.09×   |
   | (32, 7168, 2048) | 3053 (38.2%)        | 497 (6.2%)          | 6.15×   |
   | (32, 8192, 5120) | 2877 (36.0%)        | 1378 (17.2%)        | 2.09×   |
  
   Note: scale_block_k=32 rows are omitted from the 3D comparison — triton_to_mxfp8_dim1 is a 32×1
   kernel, not the 32×32 tile, so no apples-to-apples triton baseline exists for the sbk=32 path.
  
   Headline
   - 2D peaks at 47–52% of HBM3E spec peak at the largest DSV3 shapes (1.24× / 1.22× over triton
   there).
   - 2D smaller-M shapes are 1.5–3.5× over triton.
   - 3D is 1.80–6.15× over triton across every shape; biggest wins on small-K shapes where the
   per-expert triton loop pays heavy launch/rearrange overhead.
   - One parity row: 2D 32x1 (131072, 2048) lands at 0.98× (both kernels saturate HBM at large M for
   this shape).
  
   Numbers are single-run; happy to rerun with N=10 + min/median if that's useful. FlyDSL → cuTeDSL
   parity is follow-up work; this PR establishes that the AMD path beats the triton fallback.

### CODE REVIEW

**Comment by vkuzo** [comment]
@zstreet87 could you add the performance achieved (GB/s in both absolute units and % of peak) to the PR summary?

**Comment by vkuzo** [comment]
@claude review this PR

File: test/prototype/moe_training/test_kernels.py [hunk 1]
```diff
@@ -920,3 +924,358 @@ def test_cutedsl_kernels_work_with_valid_128_multiple_groups():
...
+# 3D (per-expert N-direction): N % 32, K % 256 (lane × VEC bf16 dwordx2 loads).
+# Shape grids mirror cutedsl test coverage (`(128, 8192)` × cutedsl K set
+# for 1x32; `(128, 1024)` × cutedsl K set for 32x1) wherever the FlyDSL
+# divisibility constraints above allow.
+_FLYDSL_1X32_K = (2048, 4096, 8192)
+_FLYDSL_2D_M = (1, 32, 128, 1024, 8192)
+_FLYDSL_32X1_K = (256, 512, 1536, 2048, 4096, 5120, 7168, 8192)
+_FLYDSL_3D_E = (1, 2, 4, 8)
+_FLYDSL_3D_N = (32, 64, 256)
+_FLYDSL_3D_K = (256, 1024, 4096)
```
Comment (vkuzo) [inline test/prototype/moe_training/test_kernels.py]: +1 to claude that this is a lot of cases, can we make sure this is quick to run on supported hardware

**Comment by vkuzo** [comment]
thanks for adding the benchmarks! I had one nit inline, otherwise lgtm.  Can you fix the `ruff` error?  also @danielvegamyhre should chime in as a lot of this is going into moe_training.

**Comment by zstreet87** [comment]
hey @vkuzo any movements on my PR?

File: torchao/prototype/mx_formats/config.py [hunk 2]
```diff
@@ -31,6 +31,11 @@ class MXFP8Dim1CastKernelChoice(Enum):
     CUDA = "cuda"
     TORCH = "torch"
     CUTEDSL = "cutedsl"
+    # AMD CDNA3+ counterpart of CUTEDSL via FlyDSL. FLOOR-mode scale only;
+    # all existing MXFP8TrainingRecipe entries force RCEIL today, so callers
+    # must construct an MXFP8TrainingOpConfig with scale_calculation_mode=FLOOR
+    # explicitly to actually exercise this backend until FlyDSL gains RCEIL.
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/config.py]: do you want to just do RCEIL in software?  Here is the code: https://github.com/pytorch/ao/blame/13cd013d65769db5f67cda2c3ea311cb5b70a665/torchao/prototype/mx_formats/mx_tensor.py#L168

File: test/prototype/moe_training/test_kernels.py (hunk 1 above)
Comment (vkuzo) [reply test/prototype/moe_training/test_kernels.py]: this is still a lot of cases, can we cut down?

**Comment by vkuzo** [comment]
overall it looks reasonable to me.  I would recommend just doing RCEIL in software in your path, the FLOOR method is not widely used anymore and will limit the practical usability of your kernels.  I think @danielvegamyhre should accept this one.

looks like there are also some ruff errors, could you rebase and make sure ruff passes

**Comment by zstreet87** [comment]
@vkuzo done!

**Comment by wenchenvincent** [comment]
Hi @danielvegamyhre Could you help review this PR?

**Comment by vkuzo** [comment]
looks like `ruff` is still failing, could you fix that

**Comment by zstreet87** [comment]
Hi @vkuzo all green on my side. Please let me know if there is anything else needed from my side.

**Comment by vkuzo** [comment]
@danielvegamyhre was on vacation and is now back - Daniel can you do the final review

**Comment by zstreet87** [comment]
Hi @vkuzo and @danielvegamyhre any update with the review?

**Comment by zstreet87** [comment]
gentle ping @vkuzo @danielvegamyhre

**Comment by vkuzo** [comment]
@claude review, focus on numerical correctness, consistency with the cutedsl numerics, and edge case handling

File: test/prototype/moe_training/test_kernels.py (hunk 1 above)
Comment (vkuzo) [reply test/prototype/moe_training/test_kernels.py]: how long does this take to run on AMD hardware?

**Comment by vkuzo** [comment]
sorry for the delay, I was OOO and now back! benchmarks look good.  A few questions:
1. how long do the tests take to run? Seems like a lot of cases.  Can we make sure it's O(seconds) since these are unit tests
2. the "tail-handling code has zero test coverage" comment from Claude review seems worth addressing

**Comment by zstreet87** [comment]
Thanks @vkuzo! Addressed both points in the latest push:

1. Test runtime. The FlyDSL numerics tests were full M×K×N×E cross-products (399 cases, ~57s). Collapsed each into explicit shape tuples — one full-tile shape, one edge shape, and one tail shape per kernel — while keeping the dtype × scaling_mode sweep since bf16/f32 and floor/rceil select genuinely distinct numeric paths. Now 55 cases, ~10s on MI355X (~2s of that is fixed import/collection).

2. Tail-handling coverage. Added a K-tail shape to every FlyDSL numerics grid so the per-lane TAIL_BLOCKS/HAS_TAIL paths actually run against the to_mx reference:
  - 1x32: K=2080 (% 2048 != 0)
  - 32x1 & 3D: K=1408 (DSV3 hidden_dim, % 256 != 0)
  - The config test now hits K=1408 under scale_block_k=32 and blocked_scale_output=True, covering the two tail combos the review flagged as untested.

All tail cases are bit-exact vs to_mx for both floor and rceil. Also rebased on latest main; ruff lint + format pass clean.

Re the RCEIL-NaN / FLOOR-denormal divergences from the bot review — those only trigger on non-finite or denormal-scale amax, which is out of contract for these kernels. Happy to add explicit "non-finite amax unsupported" notes if you'd prefer that documented rather than left implicit.

**Comment by vkuzo** [comment]
@claude review this again

**Review by vkuzo — APPROVED** [review]
thank you, this looks good! Going to wait for CI and merge if green

curious if you have context on the technical reason triton falls short here and we need to reach for flyDSL
